torch/csrc/autograd/python_variable.cpp

```diff
@@ -151,34 +182,73 @@ static int THPVariable_clear(THPVariable *self)
   Py_CLEAR(self->backward_hooks);
   const auto& tensor = THPVariable_Unpack(self);
   if (tensor.defined()) {
-    if (auto grad_acc = torch::autograd::impl::try_get_grad_accumulator(tensor)) {
-      grad_acc->pre_hooks().clear();
+    // Two situations to consider:
+    //    PyObject -owns-> Tensor
+    //        unsafeIsBorrowed() is FALSE.  We're obligated to look through
+    //        Tensor to break references.  Clearing cdata could induce the
+    //        destruction of the C++ Tensor.
+    //    Tensor -owns-> PyObject
+    //        unsafeIsBorrowed() is TRUE.  We're deallocating the PyObject
+    //        because Tensor asked us to (it's already destructing).
+
+    if (!self->cdata.unsafeIsBorrowed()) {
+      // TODO: empirically, this assert appears to be untrue
+      TORCH_INTERNAL_ASSERT(!tensor.unsafeGetTensorImpl()->owns_pyobj());
+      if (auto grad_acc = torch::autograd::impl::try_get_grad_accumulator(tensor)) {
+        grad_acc->pre_hooks().clear();
+      }
     }
-    // We must clear the pyobj field in the base C++ Variable, to ensure
-    // that if we attempt to pass the Variable to Python, we don't
-    // attempt to reuse the (now-dead) PyObject.
-    //
-    // One non-obvious consequence of this: if you have a tensor x, you
-    // take its id(), and then you let it become dead in Python, if you
-    // get another reference to the tensor in Python later (because you
-    // passed it from C++ to Python), you'll get a *different* id() the
-    // second time around.  So you better make sure that if you're using
-    // id() to keep track of Tensors, you better make sure their Python
-    // objects stay live, buster!  See
-    // https://github.com/pytorch/pytorch/issues/22884 for an example of
-    // this actually showing up.
-    tensor.unsafeGetTensorImpl()->unchecked_clear_pyobj(self_interpreter.get());
-  }
-  self->cdata.reset();
+    // TODO: figure out why this is necessary
+    // tensor.unsafeGetTensorImpl()->unchecked_clear_pyobj(self_interpreter.get());
+  }
+  self->cdata = MaybeOwned<Variable>();
   return 0;
 }
 
-static void THPVariable_dealloc(THPVariable* self)
-{
-  PyObject_GC_UnTrack(self);
-  THPVariable_clear(self);
-  self->cdata.~Variable();
-  Py_TYPE(self)->tp_free((PyObject*)self);
+// returns true if successfully rezzed; if so, cancel the
+// rest of deallocation
+static bool THPVariable_tryResurrect(THPVariable* self) {
+
+  const auto& tensor = THPVariable_Unpack(self);
+
+  // Is this true or not???  Triggered by TestAutograd.test_variable_traverse
```

Comment:
I'm not even sure how you can get a python binding for an undefined Tensor?
We had to resort to a custom c++ Node to be able to get one for gradcheck puposes here: https://github.com/pytorch/pytorch/blob/098d9975a7c6b33ada382bd51f322f7f5c694a17/torch/autograd/gradcheck.py#L756